### ISSUE TITLE
Feature request: YAML frontmatter in markdown preview

### ISSUE THREAD

**Issue by rnarian** [gogits/gogs#2931]
It would be nice, if markdown preview could display YAML front matter in markdown files as tables. Right now files with YAML front matter seem a little broken.

Gogs:

> ![bildschirmfoto 2016-04-05 um 18 28 20](https://cloud.githubusercontent.com/assets/1174951/14289703/46bd7010-fb5c-11e5-97c9-2af36a23a937.png)
> (https://try.gogs.io/rnarian/yaml-frontmatter/src/master/test.md)

Github:

> ![bildschirmfoto 2016-04-05 um 18 28 38](https://cloud.githubusercontent.com/assets/1174951/14289728/5ac0495c-fb5c-11e5-9aed-bf9837a8b33e.png)
> (https://github.com/rnarian/yaml-frontmatter/blob/master/test.md)

**Comment by Unknwon** [comment]
I think it's more like a feature request to https://github.com/russross/blackfriday...

**Comment by shurcooL** [comment]
I agree you may not want to implement this functionality in gogs directly, since it's pretty common and general.

At the same time, I'm not sure it falls within scope of `blackfriday`, see my https://github.com/russross/blackfriday/issues/254#issuecomment-206656510.

Isn't there some package that uses `blackfriday` and adds support for YAML headers? If such a thing exists, gogs can use it.

**Comment by rnarian** [comment]
> Isn't there some package that uses `blackfriday` and adds support for YAML headers? If such a thing exists, gogs can use it.

@Unknwon Maybe that's an option?

**Comment by Unknwon** [comment]
@rnarian 

> Maybe that's an option?

Yeah, but that's not a affirmative sentence.

**Comment by rnarian** [comment]
Alright, I see. It would be great to see this implemented in gogs sometime in the future though. Many static site generators are using the yaml front matter and when the infos are more complex than in my example the previews get more and more unreadable.

**Comment by Unknwon** [comment]
It is easy to fix the particular situation you have, which is at the very beginning of the file, but that's not somewhat ideal... So looking forward someone might find a useful package or blackfriday extension.

**Comment by mblauser** [comment]
Bumping this old issue because the landscape has changed a lot since 2016.

Google recently published a draft spec called the [Open Knowledge Format](https://github.com/GoogleCloudPlatform/knowledge-catalog/blob/main/okf/SPEC.md), which defines a knowledge base as a directory of markdown files where every file carries YAML frontmatter (a `type` field is required). I keep my knowledgebase markdown files in this style, and browsing them in Gogs is rough: the opening `---` turns into a horizontal rule or a heading, and the metadata blob gets mashed into the page content.

Suggested solutions have been proposed a couple of times already, once in this repo and once downstream.

PR #4643 (2017) parsed frontmatter with Hugo's page parser (YAML/TOML/JSON), rendered it as an HTML table, and fell back to plain rendering if parsing failed. It was closed unmerged, but the approach still looks right to me, especially the fallback behavior.

Gitea filed the same bug as go-gitea/gitea#5377 and shipped a fix in 1.12.0 (#11016, #11047). They render frontmatter as a table by default via yuin/goldmark-meta, with per-file control: `gitea: none` hides it, `table` shows it inline, `details` tucks it behind a collapsible block.

So my ask, roughly in order of importance:

1. Only treat it as frontmatter if the `---` block starts on the very first line of the file.
2. Parse the YAML separately and render the rest of the file as normal markdown.
3. Show the metadata as a small key/value table. Collapsible via `<details>` is fine, but please give the summary real text like "Metadata". Codeberg hit a bug where an empty summary renders as a tiny arrow nobody can find.
4. If full YAML parsing is too much for a first pass, even rendering the frontmatter block as a monospace code snippet would be a big visual improvement over the current mangled output. It keeps the metadata readable and clearly separated from the body without needing a table renderer.
5. If the YAML is malformed or never closed, just render it raw like today. Don't swallow content.
6. Configuration at two levels would be nice: a repository-wide setting for how frontmatter is displayed (table, code snippet, hidden), plus a per-file opt-out. Matching Gitea's `gitea: none` for the per-file side keeps documents portable between the two.

Since Gogs and Gitea share a lot of history, borrowing the goldmark-meta approach is probably less work than reviving the Hugo parser dependency from #4643, though honestly either one fixes the problem. YAML alone covers OKF, Jekyll, and Hugo's default case; TOML/JSON like #4643 had would be a bonus.

Happy to test against a real OKF-style repo if someone picks this up.
